File: rust/59.spiral-matrix-ii.rs
/*
 * @lc app=leetcode id=59 lang=rust
 *
 * [59] Spiral Matrix II
 */

// @lc code=start
impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let n = n as usize;
        let mut matrix = vec![vec![0; n]; n];
        let (mut top, mut right, mut bottom, mut left) = (0, n, n, 0);
        let mut k = 1;
        loop {
            if left == right {
                break;
            }
            for j in left..right {
                matrix[top][j] = k;
                k += 1;
            }
            top += 1;
            if top == bottom {
                break;
            }
            for i in top..bottom {
                matrix[i][right - 1] = k;
                k += 1;
            }
            right -= 1;
            if left == right {
                break;
            }
            for j in (left..right).rev() {
                matrix[bottom - 1][j] = k;
                k += 1;
            }
            bottom -= 1;
            if top == bottom {
                break;
            }
            for i in (top..bottom).rev() {
                matrix[i][left] = k;
                k += 1;
            }
            left += 1;
        }
        matrix
    }
}
// @lc code=end

File: javascript/59.spiral-matrix-ii.js
/*
 * @lc app=leetcode id=59 lang=javascript
 *
 * [59] Spiral Matrix II
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
  const matrix = new Array(n).fill(null).map(() => new Array(n).fill(0));
  let top = 0,
    right = n - 1,
    bottom = n - 1,
    left = 0;
  let k = 1;
  while (k <= n ** 2) {
    for (let j = left; j <= right; j++) matrix[top][j] = k++;
    top++;

    for (let i = top; i <= bottom; i++) matrix[i][right] = k++;
    right--;

    for (let j = right; j >= left; j--) matrix[bottom][j] = k++;
    bottom--;

    for (let i = bottom; i >= top; i--) matrix[i][left] = k++;
    left++;
  }
  return matrix;
};
// @lc code=end
